### PR TITLE
Added cleanup code such that all code coverage run tests pass.

### PR DESCRIPTION
**Pull request by ppt-adsk**
Code coverage builds have to leak retained scene indices because of a crash in the retained scene index destructor, on Clang code coverage builds alone.  This creates a maintenance burden when such scene indices also manage other resources.  This pull request adds a Destroy() method to certain classes so that proper cleanup can be done in code coverage builds, even if retained scene indices are leaked to avoid crashing.

### CODE REVIEW

File: lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraSceneIndex.cpp [hunk 1]
```diff
@@ -533,6 +533,11 @@ MayaHydraSceneIndex::MayaHydraSceneIndex(
 }
 
 MayaHydraSceneIndex::~MayaHydraSceneIndex()
+{
+    _Destroy();
+}
+
+void MayaHydraSceneIndex::_Destroy()
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraSceneIndex.cpp:540]: Add Destroy() method to perform proper cleanup on code coverage builds.

File: lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.h [hunk 2]
```diff
@@ -99,6 +99,11 @@ class MayaUsdProxyShapeSceneIndex : public HdSingleInputFilteringSceneIndexBase
         _SendPrimsDirtied(entries);
     }
 
+#ifndef CODE_COVERAGE_WORKAROUND
+private:
+#endif
+    void _Destroy();
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.h:105]: Add Destroy() method to perform proper cleanup on code coverage builds.

File: lib/mayaHydra/hydraExtensions/sceneIndex/registration.cpp [hunk 3]
```diff
@@ -306,6 +313,14 @@ class PathInterfaceSceneIndex : public Fvp::PathInterfaceSceneIndexBase
...
 
     ~PathInterfaceSceneIndex() {
+        Destroy();
+    }
+
+#ifdef CODE_COVERAGE_WORKAROUND
+    friend struct PXR_NS::MayaUsdSceneIndexRegistration;
+#endif
+
+    void Destroy() {
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/registration.cpp:323]: Add Destroy() method to perform proper cleanup on code coverage builds.

File: lib/mayaHydra/hydraExtensions/sceneIndex/registration.cpp [hunk 4]
```diff
@@ -339,6 +354,18 @@ struct MayaUsdSceneIndexRegistration : public MayaHydraSceneIndexRegistration
         auto proxyShapeSceneIndex = TfDynamic_cast<MayaUsdProxyShapeSceneIndexRefPtr>(pluginSceneIndex);
         proxyShapeSceneIndex->UpdateTime();
     }
+
+#ifdef CODE_COVERAGE_WORKAROUND
+    void Destroy() override {
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/registration.cpp:359]: Invoke Destroy() methods.

File: lib/mayaHydra/mayaPlugin/renderOverride.cpp [hunk 5]
```diff
@@ -1008,7 +1008,13 @@ void MtohRenderOverride::_SetRenderPurposeTags(const MayaHydraParams& delegatePa
 
 void MtohRenderOverride::_ClearMayaHydraSceneIndex()
 {
+#ifdef CODE_COVERAGE_WORKAROUND
+    // Leak the Maya scene index for code coverage, as its base class
+    // HdRetainedSceneIndex dtor crashes in Windows clang code coverage build.
+    _mayaHydraSceneIndex->_Destroy();
```
Comment (ppt-adsk) [inline lib/mayaHydra/mayaPlugin/renderOverride.cpp:1014]: Invoke Destroy() method.

File: lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.h (hunk 2 above)
Comment (lilike-adsk) [reply lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.h:105]: Do we have a ticket to track when to remove the CODE_COVERAGE_WORKAROUND?

Comment (ppt-adsk) [reply lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.h:105]: Just created one: HYDRA-1297

**Review by lilike-adsk — APPROVED** [review]

**Pull request merged**